lib/core.js

```diff
@@ -194,6 +194,88 @@ module.exports = class Core {
     return new this(header, compat, crypto, oplog, bigHeader, tree, blocks, bitfield, verifier, legacy, opts.onupdate || noop, opts.onconflict || noop)
   }
 
+  async quickAudit () {
+    await this._mutex.lock()
+
+    let drops = 0
+    let totalDrops = 0
+
+    try {
+      await this._flushOplog()
+
+      const c = require('compact-encoding')
+      const last = this.bitfield.findLast(true, this.header.tree.length)
+      const [offset, size] = await this.tree.byteRange(last * 2)
+      const byteLength = offset + size
+      const length = last + 1
+
+      const data = await new Promise((resolve) => {
+        this.blocks.storage.read(0, byteLength, (_, data) => resolve(data || null))
+      })
+
+      const tree = await new Promise((resolve) => {
+        this.tree.storage.read(0, length * 2 * 40 - 40, (_, data) => resolve(data || null))
+      })
+
+      let i = 0
+      let nextOffset = -1
+      while (i < length) {
+        const has = this.bitfield.get(i)
+
+        if (!has) {
+          if (i + 1 === length) break
+          i = this.bitfield.findFirst(true, i + 1)
+          if (i < 0) break
+          nextOffset = -1
+          continue
+        }
+
+        if (nextOffset === -1) {
+          nextOffset = await this.tree.byteOffset(i * 2)
+        }
+
+        const state = { start: i * 80, end: i * 80 + 40, buffer: tree }
+        const len = c.uint64.decode(state)
+        const expectedHash = c.fixed32.decode(state)
+        const blk = data.subarray(nextOffset, nextOffset + len)
+        const hash = hypercoreCrypto.data(blk)
+
+        nextOffset += blk.byteLength
+
+        if (!b4a.equals(hash, expectedHash)) {
+          const entry = {
+            userData: null,
+            treeNodes: null,
+            treeUpgrade: null,
+            bitfield: {
+              start: i,
+              length: 1,
+              drop: true
+            }
+          }
+
+          await this.oplog.append([entry], false)
+
+          this._setBitfield(i, false)
+          drops++
+          totalDrops++
+
+          if (drops >= 4) {
+            await this._flushOplog()
+            drops = 0
+          }
+        }
+
+        i++
+      }
+    } finally {
+      await this._mutex.unlock()
+    }
+
+    if (drops) await this.flush()
```

Comment:
Can anything bad happen if an operation runs between the `await this._mutex.unlock()` just above and the `this.flush()`? There's no guarantee such an operation wouldn't access the mutex.

Reply:
Couldn't the same result as calling `this.flush()` be achieved within the mutex lock, just before closing the `try`? `flush` just does

```
      this._manifestFlushed = true
      this._autoFlush = 4
      await this._flushOplog()
```
I'm not sure if the first 2 statements are relevant, so perhaps just calling `await this._flushOplog()` suffices

Reply:
it should be fine, its just not in try since it also requires the mutex.